lightning-background-processor/src/lib.rs

```diff
@@ -219,6 +224,203 @@ where A::Target: chain::Access, L::Target: Logger {
+
+					if let Err(e) = $persister.persist_graph(network_graph) {
+						log_error!($logger, "Error: Failed to persist network graph, check your disk and permissions {}", e)
+					}
+
+					last_prune_call = Instant::now();
+					have_pruned = true;
+				} else {
+					log_trace!($logger, "Not pruning network graph, either due to pending rapid gossip sync or absence of a prunable graph.");
+				}
+			}
+
+			if last_scorer_persist_call.elapsed().as_secs() > SCORER_PERSIST_TIMER {
+				if let Some(ref scorer) = $scorer {
+					log_trace!($logger, "Persisting scorer");
+					if let Err(e) = $persister.persist_scorer(&scorer) {
+						log_error!($logger, "Error: Failed to persist scorer, check your disk and permissions {}", e)
+					}
+				}
+				last_scorer_persist_call = Instant::now();
+			}
+		}
+
+		// After we exit, ensure we persist the ChannelManager one final time - this avoids
+		// some races where users quit while channel updates were in-flight, with
+		// ChannelMonitor update(s) persisted without a corresponding ChannelManager update.
+		$persister.persist_manager(&*$channel_manager)?;
+
+		// Persist Scorer on exit
+		if let Some(ref scorer) = $scorer {
+			$persister.persist_scorer(&scorer)?;
+		}
+
+		// Persist NetworkGraph on exit
+		if let Some(network_graph) = $gossip_sync.network_graph() {
+			$persister.persist_graph(network_graph)?;
+		}
+
+		Ok(())
+	} }
+}
+
+/// Processes background events in a future.
+///
+/// `sleeper` should return a future which completes in the given amount of time and returns a
+/// boolean indicating whether the background processing should continue. Once `sleeper` returns a
+/// future which outputs false, the loop will exit and this function's future will complete.
+///
+/// See [`BackgroundProcessor::start`] for information on which actions this handles.
+#[cfg(feature = "futures")]
+pub async fn process_events_async<
+	'a,
+	Signer: 'static + Sign,
+	CA: 'static + Deref + Send + Sync,
+	CF: 'static + Deref + Send + Sync,
+	CW: 'static + Deref + Send + Sync,
+	T: 'static + Deref + Send + Sync,
+	K: 'static + Deref + Send + Sync,
+	F: 'static + Deref + Send + Sync,
+	G: 'static + Deref<Target = NetworkGraph<L>> + Send + Sync,
+	L: 'static + Deref + Send + Sync,
+	P: 'static + Deref + Send + Sync,
+	Descriptor: 'static + SocketDescriptor + Send + Sync,
+	CMH: 'static + Deref + Send + Sync,
+	RMH: 'static + Deref + Send + Sync,
+	EH: 'static + EventHandler + Send,
+	PS: 'static + Deref + Send,
+	M: 'static + Deref<Target = ChainMonitor<Signer, CF, T, F, L, P>> + Send + Sync,
+	CM: 'static + Deref<Target = ChannelManager<Signer, CW, T, K, F, L>> + Send + Sync,
+	PGS: 'static + Deref<Target = P2PGossipSync<G, CA, L>> + Send + Sync,
+	RGS: 'static + Deref<Target = RapidGossipSync<G, L>> + Send,
+	UMH: 'static + Deref + Send + Sync,
+	PM: 'static + Deref<Target = PeerManager<Descriptor, CMH, RMH, L, UMH>> + Send + Sync,
+	S: 'static + Deref<Target = SC> + Send + Sync,
+	SC: WriteableScore<'a>,
```

Comment:
Wonder if there is a way to move the common type parameters to `BackgroundProcessor` struct so as not to repeat this long trait bounds list. Then have an "implementation" type parameter to choose between sync and async. Then the macro could be a function without repeating all the trait bounds again.

Reply:
I think those would need to be associated types, so the user would essentially implement the (empty) trait by defining the associated types? Maybe we'd could provided some stock implementations with partially defined type aliases? Just thinking out loud.

Reply:
Seems like it would be a bit cleaner though as `BackgroundProcessor`'s impl block wouldn't have a long list of bounds. Instead they would be associated types on the trait. Maybe could even turn the macro into a default method implementation on the trait?

Reply:
Hrmm... right. I guess if the `$loop_exit_check` and `$await` usages could be moved to the start of the loop, then the default method could instead be the loop body (taking `updates_available` and `await_time`). Would need a default method for shutdown clean-up, too. Then each version (sync and async) would be written to include a loop and calls to each method, one inside the loop and one after it.

Not sure if it's worth doing, though. And might be a reason why we process events at the start of the loop before those checks.

Reply:
The only other alternative that I've considered, but haven't tried in any sense, is to define some sort of `Executor` trait that would abstract away execution logic, which would be defined by two different implementations. And parameterize `BackgroundProcessor::start` with this where the `Executor` can take a closure (passed from `BackgroundProcessor`) to execute.

Reply:
`BackgroundProcessor::start` would pass it a closure that has captured any variables it needs. The closure would also callback into the `Executor` as needed. It's possible there's something else that wouldn't make it possible, but at very least I don't think the `Executor` would need all the type parameters.

Reply:
I was thinking that `BackgroundProcessor` would be used for both implementations, but all the thread code would be moved to an `Executor` implementation. `BackgroundProcessor::start` would be passed an `Executor` and would define the loop body as a closure capturing the other parameters. It would call some method on `Executor` passing the closure as the parameter. I suppose the closure would need to call back in the `Executor` for `$await` and `$loop_exit_check` functionality.

Reply:
I was proposing to add an `Executor` type parameter to `BackgroundProcessor::start` and have two implementations, moving all the `std::thread` code to one of them. Something like:

```rust
pub trait Executor {
        fn execute<F: Fn>(&self, closure: F);
        fn await(&self);
        fn loop_exit_check(&self) -> bool;
}

pub struct ThreadExecutor {
        stop_thread: Arc<AtomicBool>,
        thread_handle: Option<JoinHandle<Result<(), std::io::Error>>>,
}

impl ThreadExecutor { ... }

impl Executor for ThreadExecutor {
        fn execute<F: Fn>(&self, closure: F) {
                self.thread_handle = Some(thread::spawn(move || -> Result<(), std::io::Error> {
                        closure()
                });
        }

        // ...
}

struct FutureExecutor { ... }

impl FutureExecutor { ... }

impl Executor for FutureExecutor { ... }
```

`BackgroundProcessor::start` would call `execute` on the `Executor` with a closure. The closure would need to call `await` and `loop_exit_check` on the `Executor`. I'm guessing the tricky part would be getting those to functions to work with `FutureExecutor`. So maybe it wouldn't be possible.

Reply:
True. That best I could find was this: https://docs.rs/maybe-async/latest/maybe_async/

Reply:
Probably could just feature gate two different versions if they are mutually exclusive, but probably not worth doing here. Just figured we could avoid the duplicative type parameters.

Reply:
Btw. it's probably worth mentioning here that BDK went this route to provide blocking/async chain client implementations side-by-side (cf. https://github.com/bitcoindevkit/bdk/blob/master/macros/src/lib.rs#L84-L146). IIUC, they ran into some issues when splitting off the Esplora client from the main project and are looking into providing alternatives to give the user more control which impl they want to use.